docs/proposals/18-shoot-CA-rotation.md

```diff
@@ -0,0 +1,195 @@
+
+### Non-Goals
+
+- Creating a process that runs fully automated without shoot owner interaction. As the shoot owner controls some secrets that would probably not even be possible.
+- Forcing the shoot owner to rotate after a certain time period. Our goal rather is to issue long running certificates and let the user decide depending on their requirements to rotate as needed.
+
+## Proposal
+
+#### kube-api server CA rotation
+
+The propsal section includes a detailed description of all steps involved for rotating from a given `CA0` to the target `CA1`.
+
+`t0`: Today's situation
+
+- `kube-apiserver` uses SERVER CERT signed by `CA0` and trusts CLIENT CERTS signed by `CA0`
+- `kube-controller-manager` issues new CLIENT CERTS with `CA0`
+- kubeconfig trusts only `CA0`
+- `ServiceAccount` secrets trust only `CA0`
+- kubelet uses CLIENT CERT signed by `CA0`
+
+`t1`: Shoot owner triggers first step of CA rotation process (--> phase one is started):
+
+- Generate `CA1`
+- Generate `CLIENT_CA1`
+- Update `kube-apiserver`, `kube-scheduler`, etc. to trust CLIENT CERTS signed by both `CA0` and `CLIENT_CA1` (`--client-ca-file` flag)
+- Update `kube-controller-manager` to issue new CLIENT CERTS now with `CLIENT_CA1`
+- Update kubeconfig so that its CA bundle contains both `CA0` and`CA1` (if kubeconfig still contains a legacy CLIENT CERT then rotate the kubeconfig)
+- Update `kube-controller-manager`to include   both `CA0` and `CA1` in `ServiceAccount` secrets. Then update `Service Account` secrets.
+- Restart control plane components so that their CA bundle contains both `CA0` and `CA1` 
+- Renew CLIENT CERTS (sign them with `CLIENT_CA1`) for control plane components (Prometheus, DWD, legacy VPN)
+- Trigger node rollout
+  - This issues new CLIENT CERTS for all kubelets signed by `CLIENT_CA1`
+  - This restarts all `Pod`s and propagates `CA0` and `CA1` into their mounted `ServiceAccount` secrets (note CAs can not be reloaded by go client, therefore we need a restart of pods.)
+- _Ask user to exchange all their client credentials (kubeconfig, CLIENT CERTS issued by `CertificateSigningRequest`s) to trust both CA0 and CA1_
+
+`t2`: Shoot owner triggers second step of CA rotation process (--> phase two is started):
+
+​	Prerequisite: The shoot owner has guaranteed that they exchanged their client credentials and triggered step 2 via an annotation.
+
+- Renew SERVER CERTS (sign them with `CA1`) for `kube-apiserver`, etc.
+- Update `kube-apiserver`, `kube-scheduler`, etc. to trust only CLIENT CERTS signed by `CLIENT_CA1`
+- Update kubeconfig so that its CA bundle contains only `CA1`
+- Update `kube-controller-manager` to only contain CA1. `ServiceAccount` secrets created after this point will get secrets that include only `CA1`
+- Restart control plane components so that their CA bundle contains only `CA1`
+- Restart kubelets so that the CA bundle in their kubeconfigs contain only `CA1`
+- Delete `CA0`
```

Comment:
Similar as above, before this step completes, a new credentials rotation shouldn't be possible (triggering `t1`), but maybe that's "for free", because all of that happens within a reconciliation (?) and hence even if you trigger it anew, nothing happens until the current reconciliation is completed. Then again, even if (e.g. in an error case), nothing bad would happen (removing a superfluous CA), but the code should be OK to "respect that" if not otherwise guarded, i.e. while normally only one CA is set before `t1`, maybe there are cases (e.g. in an error case) where both CA from the last rotation are still around.